ipa-core/src/protocol/ipa_prf/validation_protocol/validation.rs

```diff
@@ -0,0 +1,508 @@
+    ///
+    /// ## Panics
+    /// Panics when recursion factor constant cannot be converted to `u128`
+    /// or when sending and receiving hashes over the network fails.
+    async fn generate_challenge<C: Context>(
+        &self,
+        ctx: &C,
+    ) -> (Vec<Fp61BitPrime>, Vec<Fp61BitPrime>) {
+        // exclude for first proof
+        let exclude_large = u128::try_from(LargeProofGenerator::RECURSION_FACTOR).unwrap();
+        // exclude for other proofs
+        let exclude_small = u128::try_from(SmallProofGenerator::RECURSION_FACTOR).unwrap();
+
+        // generate hashes
+        let hashes = ProofHash::generate_hash(self);
+
+        // send hashes
+        let send_future = hashes.send_hash(ctx);
+        let receive_future = ProofHash::receive_hash(ctx, hashes.len());
+
+        // receive other hashes
+        let ((), other_hashes) = try_join(send_future, receive_future).await.unwrap();
+
+        // compute challenges_from_proof_of_left_prover
+        // the prover computes hash_to_field(left,right)
+        // this is the right verifier of the party to the left,
+        // i.e. it has the right share, other is the left share
+        let mut challenges_from_proof_of_left_prover =
+            Vec::<Fp61BitPrime>::with_capacity(hashes.len());
+        challenges_from_proof_of_left_prover.push(hash_to_field(
+            &other_hashes.proof_from_prover_left_hash[0],
+            &hashes.proof_from_prover_left_hash[0],
+            exclude_large,
+        ));
+        for (left, right) in other_hashes
+            .proof_from_prover_left_hash
+            .iter()
+            .zip(hashes.proof_from_prover_left_hash.iter())
+        {
+            challenges_from_proof_of_left_prover.push(hash_to_field(left, right, exclude_small));
+        }
+
+        // compute challenges_from_proof_of_right_prover
+        // the prover computes hash_to_field(left,right)
+        // this is the left verifier of the party to the right,
+        // i.e. it has the left share, other is the right share
+        let mut challenges_from_proof_of_right_prover =
+            Vec::<Fp61BitPrime>::with_capacity(hashes.len());
+        challenges_from_proof_of_right_prover.push(hash_to_field(
+            &hashes.proof_from_prover_right_hash[0],
+            &other_hashes.proof_from_prover_right_hash[0],
+            exclude_large,
+        ));
+        for (left, right) in hashes
+            .proof_from_prover_right_hash
+            .iter()
+            .zip(other_hashes.proof_from_prover_right_hash.iter())
+        {
+            challenges_from_proof_of_right_prover.push(hash_to_field(left, right, exclude_small));
+        }
+
+        (
+            challenges_from_proof_of_left_prover,
+            challenges_from_proof_of_right_prover,
+        )
+    }
+}
+
+struct ProofHash {
+    proof_from_prover_left_hash: Vec<Hash>,
+    proof_from_prover_right_hash: Vec<Hash>,
```

Comment:
These are vectors of hashes, so it seems odd to call them "proof". I'd suggest a name like: `hashes_for_prover_left` and `hashes_for_prover_right`